### DISCUSSION TITLE
Can you make it so that when a bunch of notes are selected, whatever operation you do is applied to every note?

### DISCUSSION THREAD

**Discussion by Sy256** [helge17/tuxguitar#58] [General]
This is how older 2.0 branch worked, in this video you can see the difference when I try to adjust the duration of a group of notes, but after I made the video I also found that other things like palm muting, triplets, transposing, staccato are only applying to the last selected note. Is it possible to change that to how the other one is working?

See here and ignore my messy desktop lol. First I show the older 2.0, and then I switch to the remote computer I had this tuxguitar. 

https://youtu.be/NzJ4g9bvfo8

Thank you!

**Comment by guiv42** [comment]
Yes, a lot of operations can be applied to a selection instead of current beat. It's my intention to backport from 2.0beta these features, but for sure it will take some time (currently working on other features)

**Reply by Sy256** [reply]
Thanks! Yeah when I was testing out the new 'sound whenever you transpose a note' feature I realized the whole chord isn't getting transposed, like its working in 2.0beta, just the selected note. It would be cool if when you code this you could have a checkbox for the behavior it'll do if that's not tricky. Like checkbox to switch between the way it selects in this version right now and how it selects in 2.0beta, both are good for different situations.

**Comment by guiv42** [comment]
Implementation of selection-wide actions is in progress, please be patient this is time-consuming. These features in 2.0beta fork broke the compatibility with Android version, so it's not only back-porting but adaptation is required (see #62).
Set note duration action is available from release 2023-07-12. It corresponds to the action shown in video in the first post of this thread. Clean beat action is also implemented.

@Sy256: regarding increment/decrement semitone actions, you're right it could conflict with "play beat when note is transposed" feature implemented recently. However, I do not think a checkbox is needed. Globally, as a design rule: avoid requesting users to set config parameters, better find an implementation which fits all cases. Especially if parameter (checkbox) is "hidden" somewhere in a dialog most users won't even be aware of.
In this case I don't think it's necessary:
- if nothing is selected with click & drag, apply transposition action to current note only (i.e. note under the caret), and play sound of current beat. So, one can transpose a single note in a chord and still hear the full chord.
- if one beat is selected (possibly several notes), apply transposition to the full beat and play sound. So, one can transpose and hear a full chord.
- if more than one beat is selected, apply transposition to all notes of all beats, but do not play any sound. So, one can transpose many chords simultaneously, but silently

**Comment by guiv42** [comment]
This is going to take time: I'm intentionally slowing down this activity. The strategy of 2.0beta fork was simple for selection-wide actions: loop over all selected notes and apply action. But in some cases it doesn't make sense.
Since 3660f137b49ecce236486a02931c3c63974c8338, 'semitone down' now applies to all selected notes *but the notes already on fret 0*. Globally it can break the harmony:
- transpose -1/2 tone a Fm chord `133111`, you get an Em `022000`, that's fine
- once more, and you get `011000`, that doesn't make sense
- if you select several beats you don't even hear the result

It's even worse with "shift up/down" action: applying the same strategy can lead to deleting notes (e.g. shift down `32----` would lead to `7-----`). Besides that, 2.0beta implementation is too buggy to be reused (it sometimes shifts by more than 1 string). 

So I will rework this, with the strategy to apply action either to *all* selected notes, or to *none of them*. "Semitone down" action applied to a basic Em chord shall do nothing.

**Reply by Sy256** [reply]
Alright sounds good thanks! I'm happy with the way 2.0 beta works for selections even though there are those bugs, I just avoid shifting parts that I know will mess up because they're too close to the 0 frets. The difference in how selections from this one to the 2.0 beta work is the only thing stopping me from using this for everything instead of switching back and forth, everything else is extra. I mainly use it for quick chord transpositions, and highlighting everything to become a different duration all at once, while maintaining their positions. Would it be possible to bring those behaviors into this fork for a version as-is and then you can refine it in another version you think? Thanks! 

Also if a bunch of chords are selected to be transposed just make any sound playback just the first chord probably. Or none, as you mentioned above is good too.

**Reply by guiv42** [reply]
I'm sorry but I don't really understand. Which version of this repo are you using?
Regarding the actions "transpose semitone up/down" and "change note duration", the behaviors of 2023-07-15 version of this repo and 2.0beta fork should be exactly identical (see [releases](https://github.com/helge17/tuxguitar/releases)).
The only modification I have implemented is to restore compatibility with Android.
Could you please test the most recent version?

**Reply by Sy256** [reply]
Hey sorry I hadn't tried the latest version but I just did and it's all working great! Thank you that was fast! That's all I needed to switch fully to this version. One last difference between the way selecting is working between this version and 2.0 beta that its missing is being able to apply all effects (like palm mutes) and velocity changes to all selected notes, can you add that please? Thanks again!

**Reply by guiv42** [reply]
yes, as mentioned above this is planned but going on slowly. Currently struggling with "shift 1 string up/down" action

**Comment by guiv42** [comment]
Note, bug present for all effects in 2.0beta: when clicking on an effect in toolbar when a bunch of notes is selected, it *toggles* the effect on all notes. It should either *set* or *unset* the effect everywhere.
Just select 2 notes, 1st one with the effect and 2nd without. Click on the button effect, then effect is applied to 2nd note but not first. Impossible to set the effect on the 2 notes.
So, once again the simplistic approach to loop and apply action to all selected notes does not work properly.
It means a re-implementation is required, not a simple back-porting.

**Comment by Sy256** [comment]
Yeah I have experienced that bug in 2.0 beta, but personally I can live with it and avoid it usually, would it be possible to port over that behavior and just refine it in further versions if needed? For my purposes I just use the range select effects mostly for string switching and making a bunch of things palm muted, and secondary, velocity changes. Usually I have it set it a way to avoid those bugs but if it misses a couple notes I can just go set those manually. 

Also been away a few days but trying the latest version (7-27) now, I was on 7-21 before it seems. Thanks

**Reply by Sy256** [reply]
Trying 7-27 for mac, multiple selection string switch is working great! Thank you! What other changes should I test in it?

**Reply by Sy256** [reply]
Just noticed multiple select velocity also seems to be working and I like the way you can click each velocity to hear how it'll sound very cool, so now all that's missing from my go-tos is multi-select palm mute

**Reply by guiv42** [reply]
`would it be possible to port over that behavior and just refine it in further versions if needed?`
Well, not really. As already mentioned above, code from 2.0beta crashes the Android application, which is clearly not acceptable. And globally, it's a lot easier to do the job all at once and not in several steps.
`What other changes should I test in it?`
I'm currently working on effects, but I don't want to push the evolution until all of them are correctly implemented. It would not be understandable why some effects apply to whole selection, and some don't. If you would like to test progressively the features as I implement them, that would surely help. So, if you want I can publish them in a dedicated temporary branch, but this means you need to build from source. Could you do this?

**Reply by Sy256** [reply]
Hmmm I don't think I can build from source on mac is the problem, I had to get outside help from someone to be able to build 2.0 beta (they built it and sent it to me) and it took a few days and even after that it still has errors (i just lived with them and the synth playback not working) and I remember I couldn't use this version until you guys did whatever you did to make it standalone in the releases tab. But I was just asking if there's a log of what's changed somewhere that way I can test all you've done in the current released versions. Also why I'm having more notes now, I'm using it for everything.

Also I missed that it crashes the android version my bad

**Comment by guiv42** [comment]
Apply all effects to selection: merged (#89)
Help would be welcome for testing. A few tips, to test for each effect (x20):
- apply to 1 single note, not selected (under cursor)
- apply to multiple notes: notes in same chord, notes in same measure, notes across several measures
- apply to a bunch of notes where effect is already active for some notes (for effects with parameters: same effect present on different notes - not all - but with different parameters, then try to apply new effect parameters)
- apply to a bunch of notes, with and without cursor present on one note
- set/unset effect
- test with voice 1, voice 2
- test with toolbar, menu, shortcut (note: shortcut can be triggered even if toolbar icon and menu items are disabled, see example of 'dead note' effect which can be added even if no note is selected and toolbar icon and menu items are disabled)
- check activation of toolbar icon and menu item, depending from selection
- check ability to undo / redo (at least 1 bug from 2.0beta fixed)
- ...

and many combinations of the different tests mentioned above.
Don't forget "let ring" effect, only accessible through menu/shortcut (no icon in toolbar, see #87)
I checked very fast the absence of regression on Android version, some more detailed tests could be useful.

**Reply by Sy256** [reply]
Awesome, I see you asked for help in testing it but I don't see a release later than 08-09 and this was after that, how do I test them? Or is it just between you guys right now?

**Reply by guiv42** [reply]
it's in  [2023-08-13](https://github.com/helge17/tuxguitar/releases/tag/2023-08-13)

**Reply by Sy256** [reply]
Hello I've been testing it for a bit and everything seems to be working great no problems here! Thanks great work!

Edit: I just ran into one thing where it crashed upon trying to shift beat over but I think it was just a fluke because I can't get it to happen again.

**Reply by guiv42** [reply]
thanks for the tests
I understand you've tested that on mac.
Just for my understanding: did you have the possibility to check absence of regression in Android version?

**Comment by guiv42** [comment]
Just a suggestion: close this discussion now it's implemented?
